File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    int dx[4] = {1, 0, -1, 0};
    int dy[4] = {0, 1, 0, -1};
    bool exist(vector<vector<char>>& board, string word) {
        int n = board.size(), m = board[0].size();
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(board[i][j] == word[0]) {
                    if(dfs(i, j, board, word, 0)) return true;
                }
            }
        }
        return false;
    }
    
    bool dfs(int row, int col, vector<vector<char>>&board, string &word, int cur) {
        if(cur >= word.size()) return true;
        if(row < 0 || row >= board.size() || col < 0 || col >= board[0].size() || board[row][col] != word[cur] || board[row][col] == '.') return false;
        char ch = word[cur];
        board[row][col] = '.';
        for(int i=0; i<4; i++) {
            if(dfs(row + dx[i], col + dy[i], board, word, cur+1)) return true;
        }
        board[row][col] = ch;
        return false;
    }
};

File: 0079-word-search/0079-word-search.java
class Solution {
    int []dx = new int[]{1, 0, -1, 0};
    int []dy = new int[]{0, 1, 0, -1};
    public boolean exist(char[][] board, String word) {
        int n = board.length, m = board[0].length;
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if(board[i][j] == word.charAt(0)) {
                    if(dfs(board, word, i, j, 0)) return true;
                }
            }
        }
        return false;
    }
    private boolean dfs(char[][] board, String word, int i, int j, int cur) {
        if(cur >= word.length()) return true;
        if(i < 0 || i >= board.length || j < 0 || j >= board[0].length || board[i][j] != word.charAt(cur) || board[i][j] == '.') return false;
        char ch = board[i][j];
        board[i][j] = '.';
        for(int k=0; k<4; k++) {
            if(dfs(board, word, i+dx[k], j+dy[k], cur+1)) return true;
        }
        board[i][j] = ch;
        return false;
    }
}